components/p3a/nitro_utils/attestation.cc

```diff
@@ -0,0 +1,264 @@
+            parse_cert_options, cert_chain.get(), &cert_errors)) {
+      LOG(ERROR) << "Nitro verification: failed to parse certificate: "
+                 << cert_errors.ToDebugString();
+      return nullptr;
+    }
+  }
+
+  net::SHA256HashValue root_cert_fp =
+      net::X509Certificate::CalculateFingerprint256(
+          cert_chain->back()->cert_buffer());
+  if (root_cert_fp != kAWSRootCertFP) {
+    LOG(ERROR)
+        << "Nitro verification: root cert fp does not match AWS root cert fp";
+    return nullptr;
+  }
+
+  return cert_chain;
+}
+
+void ParseAndVerifyDocument(
+    std::unique_ptr<network::SimpleURLLoader> url_loader,
+    std::vector<uint8_t> nonce,
+    base::OnceCallback<void(scoped_refptr<net::X509Certificate>)>
+        result_callback,
+    std::unique_ptr<std::string> response_body) {
+  if (response_body == nullptr || response_body->empty()) {
+    LOG(ERROR) << "Nitro verification: no body received from server";
+    std::move(result_callback).Run(scoped_refptr<net::X509Certificate>());
+    return;
+  }
+
+  response_body->erase(
+      std::find_if(response_body->rbegin(), response_body->rend(),
+                   [](unsigned char ch) { return !std::isspace(ch); })
+          .base(),
+      response_body->end());
+  absl::optional<std::vector<uint8_t>> cose_encoded =
+      base::Base64Decode(*response_body);
+  if (!cose_encoded.has_value()) {
+    LOG(ERROR) << "Nitro verification: Failed to decode base64 document";
+    std::move(result_callback).Run(scoped_refptr<net::X509Certificate>());
+    return;
+  }
+
+  CoseSign1 cose_doc;
+
+  if (!cose_doc.DecodeFromBytes(*cose_encoded) || !cose_doc.payload.is_map()) {
+    LOG(ERROR) << "Nitro verification: Failed to decode COSE/CBOR document";
+    std::move(result_callback).Run(scoped_refptr<net::X509Certificate>());
+    return;
+  }
+
+  const cbor::Value::MapValue& cose_map = cose_doc.payload.GetMap();
+
+  if (!VerifyNonce(cose_map, nonce)) {
+    std::move(result_callback).Run(scoped_refptr<net::X509Certificate>());
+    return;
+  }
+
+  const network::mojom::URLResponseHead* response_info =
+      url_loader->ResponseInfo();
+  DCHECK(response_info);
+  if (!response_info->ssl_info.has_value() ||
+      response_info->ssl_info->cert == nullptr) {
+    LOG(ERROR) << "Nitro verification: ssl info is missing from response info";
+    std::move(result_callback).Run(scoped_refptr<net::X509Certificate>());
+    return;
+  }
+  scoped_refptr<net::X509Certificate> server_cert =
+      response_info->ssl_info->cert;
+
+  if (!VerifyUserDataKey(server_cert, cose_map)) {
+    std::move(result_callback).Run(scoped_refptr<net::X509Certificate>());
+    return;
+  }
+
+  std::unique_ptr<net::ParsedCertificateList> cert_chain =
+      ParseCertificatesAndCheckRoot(server_cert, cose_map);
+  if (cert_chain == nullptr) {
+    std::move(result_callback).Run(scoped_refptr<net::X509Certificate>());
+    return;
+  }
+
+  if (!cose_doc.Verify(*cert_chain)) {
```

Comment:
I assume all this parsing/validation in the main process is sec reviewed and approved from the rule of 2 POV @kdenhartog ?